xarray/namedarray/core.py

```diff
@@ -0,0 +1,342 @@
+# type. hopefully in the future we can narrow this down more
+T_Array = typing.TypeVar("T_Array", bound=typing.Any)
+
+
+class NamedArray:
+    __slots__ = ("_dims", "_data", "_attrs")
+
+    def __init__(
+        self, dims, data: npt.ArrayLike, attrs: dict[typing.Any, typing.Any] = None
+    ):
+        self._dims = self._parse_dimensions(dims)
+        self._data = data
+        self._attrs = attrs or {}
+
+    @property
+    def ndim(self) -> int:
+        """
+        Number of array dimensions.
+
+        See Also
+        --------
+        numpy.ndarray.ndim
+        """
+        return len(self.shape)
+
+    @property
+    def size(self) -> int:
+        """
+        Number of elements in the array.
+
+        Equal to ``np.prod(a.shape)``, i.e., the product of the array’s dimensions.
+
+        See Also
+        --------
+        numpy.ndarray.size
+        """
+        return math.prod(self.shape)
+
+    def __len__(self) -> int:
+        try:
+            return self.shape[0]
+        except IndexError:
+            raise TypeError("len() of unsized object")
+
+    @property
+    def dtype(self) -> np.dtype:
+        """
+        Data-type of the array’s elements.
+
+        See Also
+        --------
+        ndarray.dtype
+        numpy.dtype
+        """
+        return self._data.dtype
+
+    @property
+    def shape(self) -> tuple[int, ...]:
+        """
+        Tuple of array dimensions.
+
+        See Also
+        --------
+        numpy.ndarray.shape
+        """
+        return self._data.shape
+
+    @property
+    def nbytes(self) -> int:
+        """
+        Total bytes consumed by the elements of the data array.
+
+        If the underlying data array does not include ``nbytes``, estimates
+        the bytes consumed based on the ``size`` and ``dtype``.
+        """
+        if hasattr(self._data, "nbytes"):
+            return self._data.nbytes
+        else:
+            return self.size * self.dtype.itemsize
+
+    @property
+    def dims(self) -> tuple[Hashable, ...]:
+        """Tuple of dimension names with which this variable is associated."""
+        return self._dims
+
+    @dims.setter
+    def dims(self, value: str | Iterable[Hashable]) -> None:
+        self._dims = self._parse_dimensions(value)
+
+    def _parse_dimensions(self, dims: str | Iterable[Hashable]) -> tuple[Hashable, ...]:
+        dims = (dims,) if isinstance(dims, str) else tuple(dims)
+        if len(dims) != self.ndim:
+            raise ValueError(
+                f"dimensions {dims} must have the same length as the "
+                f"number of data dimensions, ndim={self.ndim}"
+            )
+        return dims
+
+    @property
+    def attrs(self) -> dict[typing.Any, typing.Any]:
+        """Dictionary of local attributes on this variable."""
+        if self._attrs is None:
+            self._attrs = {}
+        return self._attrs
+
+    @attrs.setter
+    def attrs(self, value: Mapping[typing.Any, typing.Any]) -> None:
```

Comment:
Equivalent but more succinct:
```suggestion
    def attrs(self, value: Mapping) -> None:
```